### ISSUE TITLE
🐛 [firebase_messaging] Can't use 3rd party plugins in background handler

### ISSUE THREAD

**Issue by asmodeoux** [FirebaseExtended/flutterfire#4424]
## Bug report

**Describe the bug**
None of 3rd party libraries being called in onBackgroundMessage. I use version 8.0.0-dev11.

Here is my onBackgroundMessage function:
```
Future<void> myBackgroundMessageHandler(RemoteMessage message) async {
  print('received data in background: $message');
  await Firebase.initializeApp();
}

```

It fails to init Firebase or any other package with these logs:
```
Tried to automatically register plugins with FlutterEngine (io.flutter.embedding.engine.FlutterEngine@d6e012f) but could not find and invoke the GeneratedPluginRegistrant.
I/flutter (23389): received data in background: Instance of 'RemoteMessage'
I/flutter (23389): FlutterFire Messaging: An error occurred in your background messaging handler:
I/flutter (23389): MissingPluginException(No implementation found for method Firebase#initializeCore on channel plugins.flutter.io/firebase_core)
```

When I had this error with previous versions I inited the plugins in Application.java and it worked, but now I moved to 8.0.0 and don't use Application as I have embedding v2. And, due to logs, it fails to init in onCreate.

Here is my onCreate function in MainActivity:
```
class MainActivity: FlutterActivity() {
    override fun onCreate(savedInstanceState: Bundle?, persistentState: PersistableBundle?) {
        super.onCreate(savedInstanceState, persistentState)
    }
    override fun configureFlutterEngine(@NonNull flutterEngine: FlutterEngine) {
        super.configureFlutterEngine(flutterEngine)
    }
}
```

And Generated:
```
@Keep
public final class GeneratedPluginRegistrant {
  public static void registerWith(@NonNull FlutterEngine flutterEngine) {
    ShimPluginRegistry shimPluginRegistry = new ShimPluginRegistry(flutterEngine);
      dev.gilder.tom.apple_sign_in.AppleSignInPlugin.registerWith(shimPluginRegistry.registrarFor("dev.gilder.tom.apple_sign_in.AppleSignInPlugin"));
    flutterEngine.getPlugins().add(new com.transistorsoft.flutter.backgroundfetch.BackgroundFetchPlugin());
    flutterEngine.getPlugins().add(new io.flutter.plugins.firebase.firestore.FlutterFirebaseFirestorePlugin());
      io.flutter.plugins.firebase.functions.FlutterFirebaseFunctionsPlugin.registerWith(shimPluginRegistry.registrarFor("io.flutter.plugins.firebase.functions.FlutterFirebaseFunctionsPlugin"));
    flutterEngine.getPlugins().add(new io.flutter.plugins.deviceinfo.DeviceInfoPlugin());
    flutterEngine.getPlugins().add(new io.flutter.plugins.firebaseanalytics.FirebaseAnalyticsPlugin());
    flutterEngine.getPlugins().add(new io.flutter.plugins.firebase.auth.FlutterFirebaseAuthPlugin());
      com.firebase_auth_ui.FirebaseAuthUiPlugin.registerWith(shimPluginRegistry.registrarFor("com.firebase_auth_ui.FirebaseAuthUiPlugin"));
    flutterEngine.getPlugins().add(new io.flutter.plugins.firebase.core.FlutterFirebaseCorePlugin());
    flutterEngine.getPlugins().add(new io.flutter.plugins.firebase.messaging.FlutterFirebaseMessagingPlugin());
    flutterEngine.getPlugins().add(new io.flutter.plugins.firebase.storage.FlutterFirebaseStoragePlugin());
    flutterEngine.getPlugins().add(new com.dexterous.flutterlocalnotifications.FlutterLocalNotificationsPlugin());
    flutterEngine.getPlugins().add(new io.flutter.plugins.flutter_plugin_android_lifecycle.FlutterAndroidLifecyclePlugin());
      com.example.imagegallerysaver.ImageGallerySaverPlugin.registerWith(shimPluginRegistry.registrarFor("com.example.imagegallerysaver.ImageGallerySaverPlugin"));
    flutterEngine.getPlugins().add(new io.flutter.plugins.imagepicker.ImagePickerPlugin());
    flutterEngine.getPlugins().add(new dev.britannio.in_app_review.InAppReviewPlugin());
    flutterEngine.getPlugins().add(new io.flutter.plugins.packageinfo.PackageInfoPlugin());
    flutterEngine.getPlugins().add(new io.flutter.plugins.pathprovider.PathProviderPlugin());
    flutterEngine.getPlugins().add(new com.baseflow.permissionhandler.PermissionHandlerPlugin());
    flutterEngine.getPlugins().add(new com.di1shuai.platform_device_id.PlatformDeviceIdPlugin());
      tuev.konstantin.respect24hour.Respect24HourPlugin.registerWith(shimPluginRegistry.registrarFor("tuev.konstantin.respect24hour.Respect24HourPlugin"));
    flutterEngine.getPlugins().add(new io.flutter.plugins.sharedpreferences.SharedPreferencesPlugin());
    flutterEngine.getPlugins().add(new com.tekartik.sqflite.SqflitePlugin());
    flutterEngine.getPlugins().add(new io.flutter.plugins.urllauncher.UrlLauncherPlugin());
  }
}
```

I absolutely can't use background handler because of this, so any advice would be much appreciated. I followed all the steps in every article, but can't make it works for 3 days already.

 Could you explain me what happens when background event arrives sot that I could debug these files? I will provide more data if needed.

### Steps to reproduce

1. Add firebase_messaging: ^8.0.0-dev.11 and follow the instructions [here](https://firebase.flutter.dev/docs/messaging/overview)
2. Set _FirebaseMessaging.onBackgroundMessage(myBackgroundMessageHandler);_ with the function above
3. Sent a push while app is in background
4. See the error logs

### Expected behavior

Any packages should be inited and available in background

### Flutter doctor

Run `flutter doctor` and paste the output below:

<details><summary>Click To Expand</summary>

```
[✓] Flutter (Channel stable, 1.22.4, on Mac OS X 10.15.6 19G2021 darwin-x64, locale ru-RU)
    • Flutter version 1.22.4 at /Users/asmodeoux/FlutterSDK/flutter
    • Framework revision 1aafb3a8b9 (5 weeks ago), 2020-11-13 09:59:28 -0800
    • Engine revision 2c956a31c0
    • Dart version 2.10.4

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/asmodeoux/Library/Android/sdk
    • Platform android-30, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.4.1, Build version 11E503a
    • CocoaPods version 1.10.0

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 49.0.2
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] Connected device (1 available)
    • GM1913 (mobile) • f981dce9 • android-arm64 • Android 10 (API 29)

• No issues found!
```

</details>

---

### Flutter dependencies

<details><summary>Click To Expand</summary>

```
Dart SDK 2.10.4
Flutter SDK 1.22.4
changelog 0.8.0+15

dependencies:
- apple_sign_in 0.1.0 [flutter]
- auto_size_text 2.1.0 [flutter]
- background_fetch 0.6.0 [flutter]
- cached_network_image 2.4.1 [flutter flutter_cache_manager octo_image]
- carousel_slider 2.3.1 [flutter]
- cloud_firestore 0.14.4 [flutter meta quiver firebase_core firebase_core_platform_interface cloud_firestore_platform_interface cloud_firestore_web]
- cloud_functions 0.7.2 [meta flutter firebase_core firebase_core_platform_interface cloud_functions_platform_interface cloud_functions_web]
- cupertino_icons 0.1.3
- easy_localization 2.3.3 [flutter intl shared_preferences args path flutter_localizations]
- equatable 0.6.1 [collection meta]
- firebase_analytics 6.3.0 [meta flutter firebase_core firebase_analytics_web firebase_analytics_platform_interface]
- firebase_auth 0.18.4+1 [meta firebase_core firebase_core_platform_interface firebase_auth_platform_interface firebase_auth_web flutter]
- firebase_auth_ui 1.0.1 [flutter]
- firebase_core 0.5.3 [firebase_core_platform_interface flutter quiver meta firebase_core_web]
- firebase_messaging 8.0.0-dev.11 [meta flutter firebase_core firebase_core_platform_interface firebase_messaging_platform_interface firebase_messaging_web]
- firebase_storage 5.2.0 [flutter firebase_core firebase_core_platform_interface firebase_storage_web firebase_storage_platform_interface quiver]
- flutter 0.0.0 [characters collection meta typed_data vector_math sky_engine]
- flutter_local_notifications 3.0.2 [flutter platform flutter_local_notifications_platform_interface timezone]
- flutter_plugin_android_lifecycle 1.0.11 [flutter]
- flutter_redux 0.5.4 [meta redux flutter]
- flutter_redux_dev_tools 0.3.0 [redux redux_dev_tools flutter_redux flutter]
- flutter_svg 0.19.1 [flutter meta path_drawing vector_math xml]
- http 0.12.2 [http_parser path pedantic]
- image_gallery_saver 1.6.7 [flutter]
- image_picker 0.6.7+15 [flutter flutter_plugin_android_lifecycle image_picker_platform_interface]
- in_app_review 1.0.4 [flutter in_app_review_platform_interface]
- intl 0.16.1 [path]
- mockito 4.1.3 [analyzer build code_builder collection dart_style matcher meta source_gen test_api]
- ntp 1.0.7 [universal_io]
- package_info 0.4.3+2 [flutter]
- path_provider 1.6.24 [flutter path_provider_platform_interface path_provider_macos path_provider_linux path_provider_windows]
- percent_indicator 2.1.8 [flutter]
- permission_handler 5.0.1+1 [flutter meta permission_handler_platform_interface]
- platform_device_id 0.1.3 [flutter device_info]
- respect_24_hour 0.0.2 [flutter]
- shared_preferences 0.5.12+4 [meta flutter shared_preferences_platform_interface shared_preferences_linux shared_preferences_macos shared_preferences_web shared_preferences_windows]
- status_alert 0.1.3 [flutter flare_flutter]
- table_calendar 2.3.2 [flutter intl simple_gesture_detector]
- url_launcher 5.7.10 [flutter url_launcher_platform_interface url_launcher_web url_launcher_linux url_launcher_macos url_launcher_windows]

dev dependencies:
- flutter_test 0.0.0 [flutter test_api path fake_async clock stack_trace vector_math async boolean_selector characters charcode collection matcher meta source_span stream_channel string_scanner term_glyph typed_data]

transitive dependencies:
- _fe_analyzer_shared 12.0.0 [meta]
- analyzer 0.40.6 [_fe_analyzer_shared args cli_util collection convert crypto glob meta package_config path pub_semver source_span watcher yaml]
- args 1.6.0
- async 2.5.0-nullsafety.1 [collection]
- boolean_selector 2.1.0-nullsafety.1 [source_span string_scanner]
- build 1.6.0 [analyzer async convert crypto logging meta path glob]
- built_collection 4.3.2 [collection quiver]
- built_value 7.1.0 [built_collection collection fixnum quiver]
- characters 1.1.0-nullsafety.3
- charcode 1.2.0-nullsafety.1
- cli_util 0.2.0 [path]
- clock 1.1.0-nullsafety.1
- cloud_firestore_platform_interface 2.2.1 [flutter meta collection firebase_core plugin_platform_interface]
- cloud_firestore_web 0.2.1+2 [flutter flutter_web_plugins http_parser meta firebase_core firebase_core_web cloud_firestore_platform_interface js]
- cloud_functions_platform_interface 3.0.3 [flutter meta firebase_core plugin_platform_interface]
- cloud_functions_web 3.1.1 [firebase_core firebase_core_web cloud_functions_platform_interface flutter flutter_web_plugins http_parser meta js]
- code_builder 3.5.0 [built_collection built_value collection matcher meta]
- collection 1.15.0-nullsafety.3
- convert 2.1.1 [charcode typed_data]
- crypto 2.1.5 [collection convert typed_data]
- dart_style 1.3.10 [analyzer args path source_span]
- device_info 1.0.0 [flutter device_info_platform_interface]
- device_info_platform_interface 1.0.1 [flutter meta plugin_platform_interface]
- fake_async 1.2.0-nullsafety.1 [clock collection]
- ffi 0.1.3
- file 5.2.1 [intl meta path]
- firebase 7.3.3 [http http_parser js]
- firebase_analytics_platform_interface 1.0.3 [flutter meta]
- firebase_analytics_web 0.1.1 [flutter flutter_web_plugins firebase firebase_analytics_platform_interface meta]
- firebase_auth_platform_interface 2.1.4 [flutter meta firebase_core plugin_platform_interface]
- firebase_auth_web 0.3.2+3 [flutter flutter_web_plugins meta http_parser intl firebase_core firebase_core_web firebase_auth_platform_interface js]
- firebase_core_platform_interface 2.1.0 [flutter meta plugin_platform_interface quiver]
- firebase_core_web 0.2.1+1 [firebase_core_platform_interface flutter flutter_web_plugins meta js]
- firebase_messaging_platform_interface 1.0.0-dev.7 [flutter meta firebase_core plugin_platform_interface]
- firebase_messaging_web 0.1.0-dev.2 [flutter flutter_web_plugins meta firebase_core firebase_core_web firebase_messaging_platform_interface js service_worker]
- firebase_storage_platform_interface 1.0.2 [flutter meta collection firebase_core plugin_platform_interface]
- firebase_storage_web 0.1.1+1 [async crypto firebase_core firebase_core_web firebase_storage_platform_interface flutter flutter_web_plugins http js meta]
- fixnum 0.10.11
- flare_dart 2.3.4
- flare_flutter 2.0.6 [flutter flare_dart meta]
- flutter_blurhash 0.5.0 [flutter meta]
- flutter_cache_manager 2.0.0 [flutter path_provider uuid http path sqflite pedantic clock file rxdart]
- flutter_local_notifications_platform_interface 2.0.0+1 [flutter plugin_platform_interface]
- flutter_localizations 0.0.0 [flutter intl characters collection meta path typed_data vector_math]
- flutter_web_plugins 0.0.0 [flutter characters collection meta typed_data vector_math]
- glob 1.2.0 [async collection node_io path pedantic string_scanner]
- http_parser 3.1.4 [charcode collection source_span string_scanner typed_data]
- image_picker_platform_interface 1.1.1 [flutter meta http plugin_platform_interface]
- in_app_review_platform_interface 1.0.5 [flutter url_launcher plugin_platform_interface platform]
- js 0.6.2
- logging 0.11.4
- matcher 0.12.10-nullsafety.1 [stack_trace]
- meta 1.3.0-nullsafety.3
- node_interop 1.2.1 [js]
- node_io 1.2.0 [file node_interop path]
- octo_image 0.3.0 [flutter flutter_blurhash]
- package_config 1.9.3 [path charcode]
- path 1.8.0-nullsafety.1
- path_drawing 0.4.1+1 [vector_math meta path_parsing flutter]
- path_parsing 0.1.4 [vector_math meta]
- path_provider_linux 0.0.1+2 [path xdg_directories path_provider_platform_interface flutter]
- path_provider_macos 0.0.4+6 [flutter]
- path_provider_platform_interface 1.0.4 [flutter meta platform plugin_platform_interface]
- path_provider_windows 0.0.4+3 [path_provider_platform_interface meta path flutter ffi win32]
- pedantic 1.9.2 [meta]
- permission_handler_platform_interface 2.0.1 [flutter meta plugin_platform_interface]
- petitparser 3.1.0 [meta]
- platform 2.2.1
- plugin_platform_interface 1.0.3 [meta]
- process 3.0.13 [file intl meta path platform]
- pub_semver 1.4.4 [collection]
- quiver 2.1.5 [matcher meta]
- redux 3.0.0
- redux_dev_tools 0.4.0 [redux]
- rxdart 0.24.1
- service_worker 0.2.4 [js]
- shared_preferences_linux 0.0.2+4 [file flutter meta path path_provider_linux shared_preferences_platform_interface]
- shared_preferences_macos 0.0.1+11 [shared_preferences_platform_interface flutter]
- shared_preferences_platform_interface 1.0.4 [meta flutter]
- shared_preferences_web 0.1.2+7 [shared_preferences_platform_interface flutter flutter_web_plugins meta]
- shared_preferences_windows 0.0.1+3 [shared_preferences_platform_interface flutter ffi file meta path path_provider_platform_interface path_provider_windows]
- simple_gesture_detector 0.1.5 [flutter]
- sky_engine 0.0.99
- source_gen 0.9.10+1 [analyzer async build dart_style glob meta path pedantic source_span]
- source_span 1.8.0-nullsafety.2 [charcode collection path term_glyph]
- sqflite 1.3.2+1 [flutter sqflite_common path]
- sqflite_common 1.0.2+1 [synchronized path meta]
- stack_trace 1.10.0-nullsafety.1 [path]
- stream_channel 2.1.0-nullsafety.1 [async]
- string_scanner 1.1.0-nullsafety.1 [charcode source_span]
- synchronized 2.2.0+2
- term_glyph 1.2.0-nullsafety.1
- test_api 0.2.19-nullsafety.2 [async boolean_selector collection meta path source_span stack_trace stream_channel string_scanner term_glyph matcher]
- timezone 0.5.9 [path]
- typed_data 1.3.0-nullsafety.3 [collection]
- universal_io 1.0.1 [meta zone_local]
- url_launcher_linux 0.0.1+4 [flutter]
- url_launcher_macos 0.0.1+9 [flutter]
- url_launcher_platform_interface 1.0.9 [flutter meta plugin_platform_interface]
- url_launcher_web 0.1.5+1 [url_launcher_platform_interface flutter flutter_web_plugins meta]
- url_launcher_windows 0.0.1+3 [flutter]
- uuid 2.2.2 [crypto convert]
- vector_math 2.1.0-nullsafety.3
- watcher 0.9.7+15 [async path pedantic]
- win32 1.7.4 [ffi]
- xdg_directories 0.1.2 [meta path process]
- xml 4.5.1 [collection convert meta petitparser]
- yaml 2.2.1 [charcode collection string_scanner source_span]
- zone_local 0.1.2
```

</details>

---

**Comment by markusaksli-nc** [comment]
Hi @asmodeoux 
Does the [official example](https://github.com/FirebaseExtended/flutterfire/tree/master/packages/firebase_messaging/firebase_messaging/example/lib) not work for you? It could be an issue with leftover config from your previous embedding.
Could you try seeing if it works with a fresh `flutter create` project and setting it up according to the [official guide](https://firebase.flutter.dev/docs/messaging/overview)?
Thank you

**Comment by asmodeoux** [comment]
> Hi @asmodeoux
> Does the [official example](https://github.com/FirebaseExtended/flutterfire/tree/master/packages/firebase_messaging/firebase_messaging/example/lib) not work for you? It could be an issue with leftover config from your previous embedding.
> Could you try seeing if it works with a fresh `flutter create` project and setting it up according to the [official guide](https://firebase.flutter.dev/docs/messaging/overview)?
> Thank you

Yeah, the example in a fresh project does work. How do I correctly fix the leftovers to make it work in my main project? Are there any hints and a list of files to check?

Because as I see there  are many people who face the same issue

**Comment by asmodeoux** [comment]
As soon as I moved files from my old one to the new one it stopper working again though, so I'd be glad to hear how does it work, which files does it use and how am I supposed to check that — thanks in advance!

**Comment by markusaksli-nc** [comment]
Any native code flutter engine setup (such as the previous android native code needed) should be removed. Also the gradle config would need to be updated according to https://firebase.flutter.dev/docs/migration. It's still hard to tell what exactly is broken. The important part is that this isn't an issue with the plugin itself, but with the outdated created project being incompatible.

The exact contents created by `flutter create` have likely changed quite a bit since you created the project. If the previous steps don't work, unless you have any native code or elaborate setup you don't want to repeat you could just use the same dart code in the new project (`flutter create --org com.myorg appname` to set up the same identifier).

**Comment by atrope** [comment]
I Have same issue after upgrading from 7.X to 8.X.. 
Was registering the plugins i use (shared preferences) as it was needed.
When Upgrading to 8.X, I Can't find a way to use shared preferences on Background Message.

It is saying that we should be able to use plugins without registering (We use V2) but when receiving Push and trying to get  item from Shared Preferences i am geting:

`E/flutter (26007): [ERROR:flutter/lib/ui/ui_dart_state.cc(177)] Unhandled Exception: MissingPluginException(No implementation found for method getAll on channel plugins.flutter.io/shared_preferences)
`

Can anyone share how to use plugins on background messages?

Thanks!

**Comment by atrope** [comment]
This issue should be Reopened ASAP, i think using 3rd party plugins is a must when receiving push (I Need to inform my backend for statistics, get userid , and save notification to database so we can show him later if needed)

On 7.X At Least we could register those plugins manually on 8.X theoretically it would load everything automatically but it is not loading...

**Comment by atrope** [comment]
@markusaksli-nc 
This is blocking us from updating any other firebase plugin because of the core dependency. 
Can you help us?

**Comment by markusaksli-nc** [comment]
@atrope Please open a new issue for the problem you are facing and fill out the details in the template, I'd be glad to help.

**Comment by atrope** [comment]
@markusaksli-nc it is the same problem..
I Can use shared preferences in a New project, I Replicated ALL files to our current setup and it does not work..

I Don't see the point on opening a new issue, we should concentrate here if we can because more users could have the same problem...

**Comment by atrope** [comment]
Ok, found out the problem.
@markusaksli-nc pleaase re-open this issue.

Everyone having this problem, please open your `GeneratedPluginRegistrant` and you will see flutter registering the plugins with 2 different types, the old plugins and the new ones.

What I realized and tested:

The plugins are inserted there alphabetically independent of V1 or V2 embedding.

Now, here's the Catch, as long as you include a plugin that uses the V2 Embedding AND do not include any V1 embedding BEFORE(and by before i mean alphabetically, by the plugin name) it will work.

In my case, i was trying to use `Shared Preferences`, it starts with "S", So i CAN include the "Social Share" plugin because the order of the `GeneratedPluginRegistrant` will be ok

`flutterEngine.getPlugins().add(new io.flutter.plugins.sharedpreferences.SharedPreferencesPlugin());   com.shekarmudaliyar.social_share.SocialSharePlugin.registerWith(shimPluginRegistry.registrarFor("com.shekarmudaliyar.social_share.SocialSharePlugin"));
 flutterEngine.getPlugins().add(new name.avioli.unilinks.UniLinksPlugin());
`
So i could theoretically use `SharedPreferences` but NOT  `unilinks`. 


I Can see 3 different approaches to fix this problem...

1 - Generate the GeneratedPluginRegistrant by Embedded version Order V2->V1, we would not be able to use V1 Plugins in our Background process, but it would not break Any V2 Plugin

2 - Remove or fix the plugin to use V2

3 - Check with Flutter Team why it's breaking when registering V1 (It's OK if it does not work, but it SHOULD not break the rest of the plugins)




Can you please check if this also applies to you, @asmodeoux (as i can see your generated, you have 5 plugins using  V1 and for your luck the first one is using V1, so none of the others would work :)) or any other that is reading this and having this problem?


Thanks!

**Comment by asmodeoux** [comment]
> Ok, found out the problem.
> @markusaksli-nc pleaase re-open this issue.
> 
> Everyone having this problem, please open your `GeneratedPluginRegistrant` and you will see flutter registering the plugins with 2 different types, the old plugins and the new ones.
> 
> What I realized and tested:
> 
> The plugins are inserted there alphabetically independent of V1 or V2 embedding.
> 
> Now, here's the Catch, as long as you include a plugin that uses the V2 Embedding AND do not include any V1 embedding BEFORE(and by before i mean alphabetically, by the plugin name) it will work.
> 
> In my case, i was trying to use `Shared Preferences`, it starts with "S", So i CAN include the "Social Share" plugin because the order of the `GeneratedPluginRegistrant` will be ok
> 
> `flutterEngine.getPlugins().add(new io.flutter.plugins.sharedpreferences.SharedPreferencesPlugin());   com.shekarmudaliyar.social_share.SocialSharePlugin.registerWith(shimPluginRegistry.registrarFor("com.shekarmudaliyar.social_share.SocialSharePlugin"));
>  flutterEngine.getPlugins().add(new name.avioli.unilinks.UniLinksPlugin());
> `
> So i could theoretically use `SharedPreferences` but NOT  `unilinks`. 
> 
> 
> I Can see 3 different approaches to fix this problem...
> 
> 1 - Generate the GeneratedPluginRegistrant by Embedded version Order V2->V1, we would not be able to use V1 Plugins in our Background process, but it would not break Any V2 Plugin
> 
> 2 - Remove or fix the plugin to use V2
> 
> 3 - Check with Flutter Team why it's breaking when registering V1 (It's OK if it does not work, but it SHOULD not break the rest of the plugins)
> 
> 
> 
> 
> Can you please check if this also applies to you, @asmodeoux (as i can see your generated, you have 5 plugins using  V1 and for your luck the first one is using V1, so none of the others would work :)) or any other that is reading this and having this problem?
> 
> 
> Thanks!

Yeah, that's exactly what I noticed back then, the order matters! I'll check your tips out, thank you and hope Flutter Team will look into it.

**Comment by atrope** [comment]
@asmodeoux  @markusaksli-nc 
Allright I'm back with a workaround.

What we will accomplish is to change the order of the `GeneratedPluginRegistrant` to put all V1 in the bottom.
To do this we need to alter the flutter tools responsible for generating that file it is pretty simple.

Open the file 
`<flutter-sdk-location>/packages/flutter_tools/lib/src/plugins.dart`

find the `_writeAndroidPluginRegistrant` function and add the following snippet AFTER the call to _extractPlatformMaps:

```
    androidPlugins.sort((Map<String, dynamic> left, Map<String, dynamic> right) {
      if (left["supportsEmbeddingV2"] != right["supportsEmbeddingV2"]) {
        return left["supportsEmbeddingV2"] ? -1 : 1;
      }
      return left["name"].compareTo(right["name"]);
    });
```

![Screen Shot 2021-02-03 at 15 53 18](https://user-images.githubusercontent.com/2945478/106794679-f08dee80-6637-11eb-98bd-5575f47dbc09.png)


Now we only need to rebuild flutter tools, to do that we need to remove  the flutter_tools.stamp:

`rm -rf  <flutter-sdk-location>/bin/cache/flutter_tools.stamp
`


That's it, now you can flutter pub get and the `GeneratedPluginRegistrant` will be sorted as we want and we will be able to use V2 embedded plugins without problems.

**Comment by MrCsabaToth** [comment]
> @asmodeoux @markusaksli-nc
> Allright I'm back with a workaround.
> ...

I heart this, but this is not FlutterFire issue. As you describe it it's a supposed generic Flutter issue, so we'd need to find or file a ticket in https://github.com/flutter/flutter/issues/
There's a FAQ ticket https://github.com/flutter/flutter/issues/62006 but I'm not 100% that's the best for this.

**Comment by atrope** [comment]
@MrCsabaToth I agree..
The last workaround (the one that we would need to register manually the plugins) was not a flutter fire issue also but rather a Flutter issue(that was fixed with V2 embedding) And we would still recommend that workaround.

I am sure more people will have this problem and will be faster to apply this workaround than wait for a flutter version to fix this :) so if anyone is having problems they should try this.

I also wrote on that issue because I understand it is bigger than only here.

I could theoretically make a PR to change this snippet on flutter tools, but I think we need first to confirm with someone from the team that it would be needed..
